### PR TITLE
TEP-0007: Add When Expressions Status

### PR DESCRIPTION
**Pull request by jerop**
Proposing the design of When Expressions Status and exploring the alternatives discussed in https://github.com/tektoncd/pipeline/issues/3139

Related TEP: https://github.com/tektoncd/community/pull/159

/cc @pritidesai @bobcatfish

### CODE REVIEW

File: teps/0007-conditions-beta.md [hunk 1]
```diff
@@ -348,6 +353,106 @@ spec:
...
+      Name:  skip-this-task
+      Task Spec:
+        Metadata:
+        Steps:
+          Image:  ubuntu
+          Name:   echo
+          Resources:
+          Script:  echo "Good Night!"
+      When:
+        Input:     foo
```
Comment (pritidesai) [inline teps/0007-conditions-beta.md:382]: A key difference between `input` here and `input` under the list of `Skipped Tasks` is unresloved `param` i.e. 

```
Input: $(params.param)
```

After resolving it:

```
Input: foo
```

Comment (jerop) [reply teps/0007-conditions-beta.md]: yes yes, added a variable to clarify that they are unresolved in the pipelinespec and resolved in the when expressions status

File: teps/0007-conditions-beta.md [hunk 2]
```diff
@@ -629,6 +738,317 @@ Provide a special kind of `runAfter` -- `runAfterEvenWhenSkipped` -- that users
 
 However, this would make what happens to the branch opaque to the guarded `Task` because its ordering-dependent `Tasks` could be executed or not be executed.
 
+### Status 
+
+#### Minimal Skipped 
+
+Add `Skipped Tasks` section to the `PipelineRunStatus` that contains a list of the `PipelineTaskName` of the skipped `Tasks`. The `WhenExpressions` that made the `Task` skipped can be found in the `PipelineSpec`, the `Parameter` variables used can be found in the `PipelineSpec` and the `Results` used from previous `Tasks` can be found in the relevant `TaskRun`. It may be involving for users to reverse-engineer to identify why a `Task` was skipped, but gives us the benefit of significantly reducing the `PipelineRunStatus` compared to what we currently have with `Conditions`.
```
Comment (sbwsg) [inline teps/0007-conditions-beta.md:745]: v v minor nit: `involving` -> `more involved` or `more work`?

**Review by sbwsg — COMMENTED** [review]
I left one tiny nit but otherwise this proposal looks good to me!

File: teps/0007-conditions-beta.md [hunk 3]
```diff
@@ -348,6 +353,110 @@ spec:
       - echo-file-does-not-exist
 ```
 
+### Status
+
+Add `Skipped Tasks` section to the `PipelineRunStatus` that contains a map with `PipelineTaskName` as key and resolved `WhenExpressions` as value. Thus, users can know why a particular `Task` was skipped. In addition, `TaskRuns` from guarded `Tasks` that execute because their `WhenExpressions` evaluate to true would have the resolved `WhenExpressions` included in the `TaskRun` status. 
+
```
Comment (pritidesai) [inline teps/0007-conditions-beta.md:359]: Hey @jerop, sorry for delay, this sounds good but we have it implemented as `minimal skipped` with [Skipped Tasks ](https://github.com/tektoncd/pipeline/blob/master/pkg/apis/pipeline/v1beta1/pipelinerun_types.go#L342) as a list of `SkippedTask`, should we update this section? 🤔

File: teps/0007-conditions-beta.md (hunk 2 above)
Comment (jerop) [reply teps/0007-conditions-beta.md]: updated that, english is hard 😅

File: teps/0007-conditions-beta.md (hunk 3 above)
Comment (jerop) [reply teps/0007-conditions-beta.md:359]: good catch @pritidesai, updated it 😊

**Comment by pritidesai** [comment]
thanks @jerop 
/lgtm

**Pull request merged**